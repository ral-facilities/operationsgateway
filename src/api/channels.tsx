import React from 'react';
import axios, { AxiosError } from 'axios';
import {
  FullChannelMetadata,
  FullScalarChannelMetadata,
  FunctionItem,
  isChannelMetadataImage,
  isChannelMetadataScalar,
  isChannelMetadataWaveform,
  RecordRow,
  timeChannelName,
} from '../app.types';
import {
  useQuery,
  UseQueryResult,
  UseQueryOptions,
} from '@tanstack/react-query';
import { ColumnDef, createColumnHelper } from '@tanstack/react-table';
import {
  roundNumber,
  TraceOrImageThumbnail,
} from '../table/cellRenderers/cellContentRenderers';
import { selectUrls } from '../state/slices/configSlice';
import { useAppDispatch, useAppSelector } from '../state/hooks';
import { readSciGatewayToken } from '../parseTokens';
import { AppDispatch } from '../state/store';
import { openImageWindow, openTraceWindow } from '../state/slices/windowSlice';

interface ChannelsEndpoint {
  channels: {
    [systemName: string]: Omit<FullChannelMetadata, 'systemName'>;
  };
}

// This metadata is always present in every record
export const staticChannels: { [systemName: string]: FullChannelMetadata } = {
  [timeChannelName]: {
    systemName: timeChannelName,
    name: 'Time',
    type: 'scalar',
    path: '/system',
  },
  shotnum: {
    systemName: 'shotnum',
    name: 'Shot Number',
    type: 'scalar',
    path: '/system',
  },
  activeArea: {
    systemName: 'activeArea',
    name: 'Active Area',
    type: 'scalar',
    path: '/system',
  },
  activeExperiment: {
    systemName: 'activeExperiment',
    name: 'Active Experiment',
    type: 'scalar',
    path: '/system',
  },
};

const fetchChannels = (apiUrl: string): Promise<FullChannelMetadata[]> => {
  return axios
    .get<ChannelsEndpoint>(`${apiUrl}/channels`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      const { channels } = response.data;

      if (!channels || Object.keys(channels).length === 0) return [];

      const convertedChannels: FullChannelMetadata[] = Object.entries(
        channels
      ).map(([systemName, channel]) => ({
        systemName,
        ...channel,
      }));

      return [...Object.values(staticChannels), ...convertedChannels];
    });
};

export interface ChannelSummary {
  first_date: string;
  most_recent_date: string;
  recent_sample: { [timestamp: string]: string | number }[];
}

const fetchChannelSummary = (
  apiUrl: string,
  channel: string
): Promise<ChannelSummary> => {
  return axios
    .get(`${apiUrl}/channels/summary/${channel}`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
export const useChannels = <T extends unknown = FullChannelMetadata[]>(
  options?: UseQueryOptions<FullChannelMetadata[], AxiosError, T, string[]>
): UseQueryResult<T, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery(
    ['channels'],
    (params) => {
      return fetchChannels(apiUrl);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      ...(options ?? {}),
    }
  );
};

export const useChannelSummary = (
  channel: string | undefined
): UseQueryResult<ChannelSummary, AxiosError> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const dataChannel =
    typeof channel !== 'undefined' && !(channel in staticChannels)
      ? channel
      : '';

  return useQuery(
    ['channelSummary', dataChannel],
    (params) => {
      return fetchChannelSummary(apiUrl, dataChannel);
    },
    {
      onError: (error) => {
        console.log('Got error ' + error.message);
      },
      enabled: dataChannel.length !== 0,
    }
  );
};

export const constructColumnDefs = (
  channels: FullChannelMetadata[],
  dispatch: AppDispatch
): ColumnDef<RecordRow>[] => {
  const columnHelper = createColumnHelper<RecordRow>();
  const myColumnDefs: ColumnDef<RecordRow>[] = [];

  channels.forEach((channel: FullChannelMetadata) => {
    const newColumnDef = columnHelper.accessor(channel.systemName, {
      id: channel.systemName,
      header: () => {
        const headerName = channel.name ? channel.name : channel.systemName;
        // Provide an actual header here when we have it
        // TODO: do we need to split on things other than underscore?
        const parts = headerName.split('_');
        const wordWrap = parts.map(
          (part, i) =>
            // \u200B renders a zero-width space character
            // which allows line-break but isn't visible
            part + (i < parts.length - 1 ? '_\u200B' : '')
        );
        return <React.Fragment>{wordWrap.join('')}</React.Fragment>;
      },
      meta: { channelInfo: channel },
      cell: isChannelMetadataScalar(channel)
        ? ({ getValue }) => {
            const value = getValue();
            return typeof value === 'number' &&
              typeof channel.precision === 'number' ? (
              <React.Fragment>
                {roundNumber(value, channel.precision, channel.notation)}
              </React.Fragment>
            ) : (
              <React.Fragment>{String(value ?? '')}</React.Fragment>
            );
          }
        : isChannelMetadataWaveform(channel)
        ? ({ row, getValue }) => {
            const value = getValue<string>();
            return (
              <TraceOrImageThumbnail
                base64Data={value}
                alt={`${channel.name ?? channel.systemName} ${
                  channel.type
                } for timestamp ${row.getValue(timeChannelName)}`}
                onClick={() =>
                  dispatch(
                    openTraceWindow({
                      recordId: (row.original as RecordRow)['_id'],
                      channelName: channel.systemName,
                    })
                  )
                }
              />
            );
          }
        : isChannelMetadataImage(channel)
        ? ({ row, getValue }) => {
            const value = getValue<string>();
            return (
              <TraceOrImageThumbnail
                base64Data={value}
                alt={`${channel.name ?? channel.systemName} ${
                  channel.type
                } for timestamp ${row.getValue(timeChannelName)}`}
                onClick={() =>
                  dispatch(
                    openImageWindow({
                      recordId: (row.original as RecordRow)['_id'],
                      channelName: channel.systemName,
                    })
                  )
                }
              />
            );
          }
        : undefined,
    });

    myColumnDefs.push(newColumnDef);
  });
  return myColumnDefs;
};

export const constructColumnDefsFromFunctions = (
  functions: FunctionItem[]
): ColumnDef<RecordRow>[] => {
  const columnHelper = createColumnHelper<RecordRow>();
  const myColumnDefs: ColumnDef<RecordRow>[] = [];

  functions.forEach((functionItem: FunctionItem) => {
    const newColumnDef = columnHelper.accessor(functionItem.name, {
      id: functionItem.name,
      header: () => <React.Fragment>{functionItem.name}</React.Fragment>,
    });

    myColumnDefs.push(newColumnDef);
  });
  return myColumnDefs;
};

export const getScalarChannels = (
  channels: FullChannelMetadata[]
): FullScalarChannelMetadata[] => {
  return channels.filter(
    (channel) => channel.type === 'scalar'
  ) as FullScalarChannelMetadata[];
};

const useScalarChannelsOptions = {
  select: (data: FullChannelMetadata[]) => getScalarChannels(data),
};

export const useScalarChannels = (): UseQueryResult<
  FullScalarChannelMetadata[],
  AxiosError
> => {
  return useChannels(useScalarChannelsOptions);
};

export const useAvailableColumns = (): UseQueryResult<
  ColumnDef<RecordRow>[],
  AxiosError
> => {
  const dispatch = useAppDispatch();
  const selectFn = React.useCallback(
    (data: FullChannelMetadata[]) => constructColumnDefs(data, dispatch),
    [dispatch]
  );
  return useChannels({ select: selectFn });
};
