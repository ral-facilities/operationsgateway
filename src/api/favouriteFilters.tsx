import {
  useMutation,
  UseMutationResult,
  useQuery,
  useQueryClient,
  UseQueryResult,
} from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { FavouriteFilter, FavouriteFilterPost } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { useAppSelector } from '../state/hooks';
import { selectUrls } from '../state/slices/configSlice';

const addFavouriteFilter = (
  apiUrl: string,
  favouriteFilter: FavouriteFilterPost
): Promise<string> => {
  const queryParams = new URLSearchParams();

  queryParams.append('name', favouriteFilter.name);
  queryParams.append('filter', favouriteFilter.filter);

  return axios
    .post<string>(
      `${apiUrl}/users/filters`,
      {},
      {
        params: queryParams,
        headers: {
          Authorization: `Bearer ${readSciGatewayToken()}`,
        },
      }
    )
    .then((response) => response.data);
};

export const useAddFavouriteFilter = (): UseMutationResult<
  string,
  AxiosError,
  FavouriteFilterPost
> => {
  const { apiUrl } = useAppSelector(selectUrls);
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (favouriteFilter: FavouriteFilterPost) =>
      addFavouriteFilter(apiUrl, favouriteFilter),
    onError: (error) => {
      console.log('Got error ' + error.message);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['favouriteFilters'] });
    },
  });
};

const fetchFavouriteFilters = (apiUrl: string): Promise<FavouriteFilter[]> => {
  return axios
    .get(`${apiUrl}/users/filters`, {
      headers: {
        Authorization: `Bearer ${readSciGatewayToken()}`,
      },
    })
    .then((response) => {
      return response.data;
    });
};

export const useFavouriteFilters = (): UseQueryResult<
  FavouriteFilter[],
  AxiosError
> => {
  const { apiUrl } = useAppSelector(selectUrls);

  return useQuery({
    queryKey: ['favouriteFilters'],

    queryFn: () => {
      return fetchFavouriteFilters(apiUrl);
    },
  });
};
