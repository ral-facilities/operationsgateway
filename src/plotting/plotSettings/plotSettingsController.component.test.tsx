import { render, screen } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { staticChannels } from '../../api/channels';
import { FullScalarChannelMetadata } from '../../app.types';
import { testScalarChannels } from '../../testUtils';
import { ChartTypeButtonsProps } from './chartTypeButtons.component';
import { PlotSettingsControllerProps } from './plotSettingsController.component';
import { PlotSettingsTextFieldProps } from './plotSettingsTextField.component';
import { XAxisTabProps } from './xAxisTab.component';
import { YAxisTabProps } from './yAxisTab.component';

describe('Plot Settings component', () => {
  let props: PlotSettingsControllerProps;
  let user: UserEvent;
  const changePlotTitle = vi.fn();
  const changePlotType = vi.fn();
  const changeXAxis = vi.fn();
  const changeXAxisScale = vi.fn();
  const changeLeftYAxisScale = vi.fn();
  const changeRightYAxisScale = vi.fn();
  const changeSelectedPlotChannels = vi.fn();
  const changeXMinimum = vi.fn();
  const changeXMaximum = vi.fn();
  const changeLeftYAxisMinimum = vi.fn();
  const changeLeftYAxisMaximum = vi.fn();
  const changeRightYAxisMinimum = vi.fn();
  const changeRightYAxisMaximum = vi.fn();
  const changeSelectedColours = vi.fn();
  const changeRemainingColours = vi.fn();

  const createView = async () => {
    // need to import like this in order for the doMock's to work
    const PlotSettingsController = (
      await vi.importActual('./plotSettingsController.component')
    ).default;
    // @ts-expect-error Type not known as not importing normally
    return render(<PlotSettingsController {...props} />);
  };

  beforeEach(() => {
    props = {
      selectedRecordTableChannels: [
        staticChannels['timestamp'] as FullScalarChannelMetadata,
      ],
      allChannels: testScalarChannels,
      plotTitle: 'Initial title',
      changePlotTitle,
      plotType: 'scatter',
      changePlotType,
      XAxis: 'test x-axis',
      changeXAxis,
      XAxisScale: 'linear',
      changeXAxisScale,
      leftYAxisScale: 'logarithmic',
      rightYAxisScale: 'linear',
      changeLeftYAxisScale,
      changeRightYAxisScale,
      leftYAxisLabel: 'left y axis label',
      rightYAxisLabel: 'right y axis label',
      changeLeftYAxisLabel: vi.fn(),
      changeRightYAxisLabel: vi.fn(),
      selectedPlotChannels: [],
      changeSelectedPlotChannels,
      changeXMinimum,
      changeXMaximum,
      changeLeftYAxisMinimum,
      changeLeftYAxisMaximum,
      changeRightYAxisMinimum,
      changeRightYAxisMaximum,
      selectedColours: [],
      remainingColours: [],
      changeSelectedColours,
      changeRemainingColours,
    };

    user = userEvent.setup({ delay: null });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('snapshots', () => {
    beforeAll(() => {
      vi.resetModules();
      vi.doMock('./plotSettingsTextField.component', () => {
        return {
          default: (props: PlotSettingsTextFieldProps) => (
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            <mock-plotSettingsTextField data-testid="mock-plotSettingsTextField">
              {Object.entries(props).map(
                ([propName, propValue]) =>
                  `${propName}=${JSON.stringify(propValue, null, 2)}\n`
              )}
              {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
              {/* @ts-ignore */}
            </mock-plotSettingsTextField>
          ),
        };
      });

      vi.doMock('./chartTypeButtons.component', () => {
        return {
          default: (props: ChartTypeButtonsProps) => (
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            <mock-chartTypeButtons data-testid="mock-chartTypeButtons">
              {Object.entries(props).map(
                ([propName, propValue]) =>
                  `${propName}=${JSON.stringify(propValue, null, 2)}\n`
              )}
              {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
              {/* @ts-ignore */}
            </mock-chartTypeButtons>
          ),
        };
      });

      vi.doMock('./xAxisTab.component', () => {
        return {
          default: (props: XAxisTabProps) => (
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            <mock-xAxisTab data-testid="mock-xAxisTab">
              {Object.entries(props).map(
                ([propName, propValue]) =>
                  `${propName}=${JSON.stringify(propValue, null, 2)}\n`
              )}
              {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
              {/* @ts-ignore */}
            </mock-xAxisTab>
          ),
        };
      });

      vi.doMock('./yAxisTab.component', () => {
        return {
          default: (props: YAxisTabProps) => (
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            <mock-yAxisTab data-testid="mock-yAxisTab">
              {Object.entries(props).map(
                ([propName, propValue]) =>
                  `${propName}=${JSON.stringify(propValue, null, 2)}\n`
              )}
              {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
              {/* @ts-ignore */}
            </mock-yAxisTab>
          ),
        };
      });
    });

    afterAll(() => {
      vi.resetModules();
      vi.doUnmock('./plotSettingsTextField.component');
      vi.doUnmock('./chartTypeButtons.component');
      vi.doUnmock('./xAxisTab.component');
      vi.doUnmock('./yAxisTab.component');
      vi.resetModules();
    });

    it('renders plot settings form correctly (timeseries plot)', async () => {
      props.XAxis = 'timestamp';
      props.XAxisScale = 'time';
      const view = await createView();

      expect(view.asFragment()).toMatchSnapshot();
    });

    it('renders plot settings form correctly (x-axis tab selected)', async () => {
      const view = await createView();

      expect(view.asFragment()).toMatchSnapshot();
    });

    it('renders plot settings form correctly (y-axis tab selected)', async () => {
      const view = await createView();
      await user.click(screen.getByRole('tab', { name: 'Y' }));

      expect(view.asFragment()).toMatchSnapshot();
    });
  });

  it('lets user switch between X and Y settings tabs', async () => {
    await createView();

    // should load X tab initially
    expect(screen.getByRole('tab', { selected: true })).toHaveTextContent('X');
    expect(screen.getByRole('tabpanel', { name: 'X' })).toBeVisible();
    expect(
      screen.queryByRole('tabpanel', { name: 'Y' })
    ).not.toBeInTheDocument();

    await user.click(screen.getByRole('tab', { name: 'Y' }));

    expect(screen.getByRole('tab', { selected: true })).toHaveTextContent('Y');
    expect(screen.getByRole('tabpanel', { name: 'Y' })).toBeVisible();
    expect(
      screen.queryByRole('tabpanel', { name: 'X' })
    ).not.toBeInTheDocument();
  });

  it('sets the correct values when plot variant changed from xy to timeseries', async () => {
    await createView();

    await user.click(screen.getByRole('button', { name: 'Timeseries' }));

    expect(changeXAxis).toHaveBeenCalledWith('timestamp');
    expect(changeXAxisScale).toHaveBeenCalledWith('time');
    expect(changeXMinimum).toHaveBeenCalledWith(undefined);
    expect(changeXMaximum).toHaveBeenCalledWith(undefined);
  });

  it('sets the correct values when plot variant changed from timeseries to xy', async () => {
    props.XAxis = 'timestamp';
    props.XAxisScale = 'time';
    await createView();

    await user.click(screen.getByRole('button', { name: 'XY' }));

    expect(changeXAxis).toHaveBeenCalledWith(undefined);
    expect(changeXAxisScale).toHaveBeenCalledWith('linear');
    expect(changeXMinimum).toHaveBeenCalledWith(undefined);
    expect(changeXMaximum).toHaveBeenCalledWith(undefined);
  });
});
