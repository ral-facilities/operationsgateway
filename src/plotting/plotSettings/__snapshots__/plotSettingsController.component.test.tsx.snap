// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plot Settings component > snapshots > renders plot settings form correctly (timeseries plot) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-plotsettingstextfield
        data-testid="mock-plotSettingsTextField"
      >
        label="Title"
value="Initial title"
onChange=undefined

      </mock-plotsettingstextfield>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-charttypebuttons
        data-testid="mock-chartTypeButtons"
      >
        plotType="scatter"
changePlotType=undefined
XAxis="timestamp"
changeXAxis=undefined

      </mock-charttypebuttons>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-9axob7-MuiPaper-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          X Axis Config
        </p>
        <mock-xaxistab
          data-testid="mock-xAxisTab"
        >
          allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "CHANNEL_ABCDE",
    "name": "Channel_ABCDE",
    "path": "/Channels/1",
    "type": "scalar"
  },
  {
    "systemName": "CHANNEL_DEFGH",
    "name": "Channel_DEFGH",
    "path": "/Channels/2",
    "type": "scalar",
    "notation": "scientific",
    "precision": 2
  },
  {
    "systemName": "CHANNEL_GHIJK",
    "name": "Channel_GHIJK",
    "path": "/Channels/3",
    "type": "scalar"
  }
]
XAxisScale="time"
XAxis="timestamp"
changeXAxis=undefined
changeXAxisScale=undefined
initialXMinimum=undefined
initialXMaximum=undefined
changeXMinimum=undefined
changeXMaximum=undefined

        </mock-xaxistab>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-ibczwg-MuiPaper-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Y Axes Config
        </p>
        <mock-yaxistab
          data-testid="mock-yAxisTab"
        >
          selectedRecordTableChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  }
]
allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "CHANNEL_ABCDE",
    "name": "Channel_ABCDE",
    "path": "/Channels/1",
    "type": "scalar"
  },
  {
    "systemName": "CHANNEL_DEFGH",
    "name": "Channel_DEFGH",
    "path": "/Channels/2",
    "type": "scalar",
    "notation": "scientific",
    "precision": 2
  },
  {
    "systemName": "CHANNEL_GHIJK",
    "name": "Channel_GHIJK",
    "path": "/Channels/3",
    "type": "scalar"
  }
]
selectedPlotChannels=[]
changeSelectedPlotChannels=undefined
initialLeftYAxisMinimum=undefined
initialLeftYAxisMaximum=undefined
changeLeftYAxisMinimum=undefined
changeLeftYAxisMaximum=undefined
initialRightYAxisMinimum=undefined
initialRightYAxisMaximum=undefined
changeRightYAxisMinimum=undefined
changeRightYAxisMaximum=undefined
leftYAxisScale="logarithmic"
changeLeftYAxisScale=undefined
rightYAxisScale="linear"
changeRightYAxisScale=undefined
leftYAxisLabel="left y axis label"
changeLeftYAxisLabel=undefined
rightYAxisLabel="right y axis label"
changeRightYAxisLabel=undefined
initialSelectedColours=[]
initialRemainingColours=[]
changeSelectedColours=undefined
changeRemainingColours=undefined

        </mock-yaxistab>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Plot Settings component > snapshots > renders plot settings form correctly (x-axis tab selected) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-plotsettingstextfield
        data-testid="mock-plotSettingsTextField"
      >
        label="Title"
value="Initial title"
onChange=undefined

      </mock-plotsettingstextfield>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-charttypebuttons
        data-testid="mock-chartTypeButtons"
      >
        plotType="scatter"
changePlotType=undefined
XAxis="test x-axis"
changeXAxis=undefined

      </mock-charttypebuttons>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-1gsv261"
      >
        <div
          class="MuiTabs-root css-1yb2e4d-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="tabs"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="X-tabpanel"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="X-tab"
                role="tab"
                tabindex="0"
                type="button"
              >
                X
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Y-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="Y-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                Y
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="X-tab"
        id="X-tabpanel"
        role="tabpanel"
      >
        <div
          class="MuiBox-root css-0"
        >
          <mock-xaxistab
            data-testid="mock-xAxisTab"
          >
            allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "CHANNEL_ABCDE",
    "name": "Channel_ABCDE",
    "path": "/Channels/1",
    "type": "scalar"
  },
  {
    "systemName": "CHANNEL_DEFGH",
    "name": "Channel_DEFGH",
    "path": "/Channels/2",
    "type": "scalar",
    "notation": "scientific",
    "precision": 2
  },
  {
    "systemName": "CHANNEL_GHIJK",
    "name": "Channel_GHIJK",
    "path": "/Channels/3",
    "type": "scalar"
  }
]
XAxisScale="linear"
XAxis="test x-axis"
changeXAxis=undefined
changeXAxisScale=undefined
initialXMinimum=undefined
initialXMaximum=undefined
changeXMinimum=undefined
changeXMaximum=undefined

          </mock-xaxistab>
        </div>
      </div>
      <div
        aria-labelledby="Y-tab"
        hidden=""
        id="Y-tabpanel"
        role="tabpanel"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Plot Settings component > snapshots > renders plot settings form correctly (y-axis tab selected) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-plotsettingstextfield
        data-testid="mock-plotSettingsTextField"
      >
        label="Title"
value="Initial title"
onChange=undefined

      </mock-plotsettingstextfield>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <mock-charttypebuttons
        data-testid="mock-chartTypeButtons"
      >
        plotType="scatter"
changePlotType=undefined
XAxis="test x-axis"
changeXAxis=undefined

      </mock-charttypebuttons>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-1gsv261"
      >
        <div
          class="MuiTabs-root css-1yb2e4d-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="tabs"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="X-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="X-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                X
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Y-tabpanel"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-ejfd77-MuiButtonBase-root-MuiTab-root"
                id="Y-tab"
                role="tab"
                tabindex="0"
                type="button"
              >
                Y
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                >
                  <span
                    class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
                    style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
                  >
                    <span
                      class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
                    />
                  </span>
                </span>
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="X-tab"
        hidden=""
        id="X-tabpanel"
        role="tabpanel"
      />
      <div
        aria-labelledby="Y-tab"
        id="Y-tabpanel"
        role="tabpanel"
      >
        <div
          class="MuiBox-root css-0"
        >
          <mock-yaxistab
            data-testid="mock-yAxisTab"
          >
            selectedRecordTableChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  }
]
allChannels=[
  {
    "systemName": "timestamp",
    "name": "Time",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "shotnum",
    "name": "Shot Number",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeArea",
    "name": "Active Area",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "activeExperiment",
    "name": "Active Experiment",
    "type": "scalar",
    "path": "/system"
  },
  {
    "systemName": "CHANNEL_ABCDE",
    "name": "Channel_ABCDE",
    "path": "/Channels/1",
    "type": "scalar"
  },
  {
    "systemName": "CHANNEL_DEFGH",
    "name": "Channel_DEFGH",
    "path": "/Channels/2",
    "type": "scalar",
    "notation": "scientific",
    "precision": 2
  },
  {
    "systemName": "CHANNEL_GHIJK",
    "name": "Channel_GHIJK",
    "path": "/Channels/3",
    "type": "scalar"
  }
]
selectedPlotChannels=[]
changeSelectedPlotChannels=undefined
initialLeftYAxisMinimum=undefined
initialLeftYAxisMaximum=undefined
changeLeftYAxisMinimum=undefined
changeLeftYAxisMaximum=undefined
initialRightYAxisMinimum=undefined
initialRightYAxisMaximum=undefined
changeRightYAxisMinimum=undefined
changeRightYAxisMaximum=undefined
leftYAxisScale="logarithmic"
changeLeftYAxisScale=undefined
rightYAxisScale="linear"
changeRightYAxisScale=undefined
leftYAxisLabel="left y axis label"
changeLeftYAxisLabel=undefined
rightYAxisLabel="right y axis label"
changeRightYAxisLabel=undefined
initialSelectedColours=[]
initialRemainingColours=[]
changeSelectedColours=undefined
changeRemainingColours=undefined

          </mock-yaxistab>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
